kornia/contrib/sam/architecture/image_encoder.py

```diff
@@ -0,0 +1,361 @@
+
+        self.norm2 = norm_layer(dim)
+        self.mlp = MLPBlock(embedding_dim=dim, mlp_dim=int(dim * mlp_ratio), act=act_layer)
+
+        self.window_size = window_size
+
+    def forward(self, x: Tensor) -> Tensor:
+        shortcut = x
+        x = self.norm1(x)
+        # Window partition
+        if self.window_size > 0:
+            H, W = x.shape[1], x.shape[2]
+            x, pad_hw = window_partition(x, self.window_size)
+
+        x = self.attn(x)
+        # Reverse window partition
+        if self.window_size > 0:
+            x = window_unpartition(x, self.window_size, pad_hw, (H, W))
+
+        x = shortcut + x
+        x = x + self.mlp(self.norm2(x))
+
+        return x
+
+
+class Attention(Module):
+    """Multi-head Attention block with relative position embeddings."""
+
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int = 8,
+        qkv_bias: bool = True,
+        use_rel_pos: bool = False,
+        rel_pos_zero_init: bool = True,
+        input_size: tuple[int, int] | None = None,
+    ) -> None:
+        """
+        Args:
+            dim: Number of input channels.
+            num_heads: Number of attention heads.
+            qkv_bias:  If True, add a learnable bias to query, key, value.
+            rel_pos: If True, add relative positional embeddings to the attention map.
+            rel_pos_zero_init: If True, zero initialize relative positional parameters.
+            input_size: Input resolution for calculating the relative positional parameter size.
+        """
+        super().__init__()
+        self.num_heads = num_heads
+        head_dim = dim // num_heads
+        self.scale = head_dim**-0.5
+
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.proj = nn.Linear(dim, dim)
+
+        self.use_rel_pos = use_rel_pos
+        if self.use_rel_pos and isinstance(input_size, tuple):
+            # initialize relative positional embeddings
+            self.rel_pos_h = nn.Parameter(zeros(2 * input_size[0] - 1, head_dim))
+            self.rel_pos_w = nn.Parameter(zeros(2 * input_size[1] - 1, head_dim))
+
+    def forward(self, x: Tensor) -> Tensor:
+        B, H, W, _ = x.shape
+        # qkv with shape (3, B, nHead, H * W, C)
+        qkv = self.qkv(x).reshape(B, H * W, 3, self.num_heads, -1).permute(2, 0, 3, 1, 4)
+        # q, k, v with shape (B * nHead, H * W, C)
+        q, k, v = qkv.reshape(3, B * self.num_heads, H * W, -1).unbind(0)
+
+        attn = (q * self.scale) @ k.transpose(-2, -1)
+
+        if self.use_rel_pos:
+            attn = add_decomposed_rel_pos(attn, q, self.rel_pos_h, self.rel_pos_w, (H, W), (H, W))
+
+        attn = attn.softmax(dim=-1)
+        x = (attn @ v).view(B, self.num_heads, H, W, -1).permute(0, 2, 3, 1, 4).reshape(B, H, W, -1)
+        x = self.proj(x)
+
+        return x
+
+
+def window_partition(x: Tensor, window_size: int) -> tuple[Tensor, tuple[int, int]]:
+    """Partition into non-overlapping windows with padding if needed.
```

Comment:
Maybe here use extract tensor patches ? Or an improved version 
https://kornia.readthedocs.io/en/latest/contrib.html#kornia.contrib.extract_tensor_patches

Reply:
i did not know about the `extract_tensor_patches`, I testest it here, and OOM, maybe because kornia version is with unfold?

Reply:
Maybe we should upgrade later with this implementation?

Reply:
sure, makes sense to me